runtimelib/src/jupyter/client.rs

```diff
@@ -1,104 +1,155 @@
     pub iopub_port: u16,
     pub stdin_port: u16,
     pub control_port: u16,
     pub hb_port: u16,
+    pub key: String,
+    pub signature_scheme: String,
     pub kernel_name: String,
-    pub ip: String,
-    key: String,
-    pub transport: String, // TODO: Enumify with tcp, ipc
-    signature_scheme: String,
-    // We'll track the connection file path here as well
-    #[serde(default)]
-    pub connection_file: String,
-    #[serde(default)]
-    pub state: String, // TODO: Use an enum
+}
+
+impl ConnectionInfo {
+    /// Read a connection file from disk and parse it into a ConnectionInfo object
+    pub async fn from_path(connection_file_path: &std::path::PathBuf) -> Result<ConnectionInfo> {
+        let content = fs::read_to_string(&connection_file_path)
+            .await
+            .unwrap_or_default();
+
+        serde_json::from_str::<ConnectionInfo>(&content).context("Failed to parse connection file")
+    }
+
+    /// format the iopub url for a ZeroMQ connection
+    pub fn iopub_url(&self) -> String {
+        format!("{}://{}:{}", self.transport, self.ip, self.iopub_port)
+    }
+
+    /// format the shell url for a ZeroMQ connection
+    pub fn shell_url(&self) -> String {
+        format!("{}://{}:{}", self.transport, self.ip, self.shell_port)
+    }
+
+    /// format the stdin url for a ZeroMQ connection
+    pub fn stdin_url(&self) -> String {
+        format!("{}://{}:{}", self.transport, self.ip, self.stdin_port)
+    }
+
+    /// format the control url for a ZeroMQ connection
+    pub fn control_url(&self) -> String {
+        format!("{}://{}:{}", self.transport, self.ip, self.control_port)
+    }
+
+    /// format the heartbeat url for a ZeroMQ connection
+    pub fn hb_url(&self) -> String {
+        format!("{}://{}:{}", self.transport, self.ip, self.hb_port)
+    }
+}
+
+/// A Jupyter runtime, representing the state of a running kernel
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct JupyterRuntime {
+    pub connection_info: ConnectionInfo,
+    pub connection_file: PathBuf,
+    pub id: Uuid,
+    // TODO: create an enum for activity state
+    pub state: String,
     pub kernel_info: Option<KernelInfoReply>,
 }
 
 impl JupyterRuntime {
-    pub async fn attach(&self) -> Result<JupyterClient, Error> {
+    /// Create a new JupyterRuntime from an on-disk connection file
+    pub async fn from_path(connection_file: PathBuf) -> Result<Self> {
+        let connection_info = ConnectionInfo::from_path(&connection_file).await?;
+        Ok(Self::new(connection_info, connection_file))
+    }
+
+    /// Create a new JupyterRuntime from a connection file and an existing ConnectionInfo object
+    /// This does not read from the connection_file path, but assumes that the ConnectionInfo
+    /// object was read from it already.
+    pub fn new(connection_info: ConnectionInfo, connection_file: PathBuf) -> Self {
+        // TODO evaluate UUID generation, should we also use connection info contents?
```

Comment:
For the moment I think this is ok as when we start to use our own launching process we can make unique IDs within the connection file name so that they're always new.